flowmachine/flowmachine/core/server/action_handlers.py

```diff
@@ -0,0 +1,327 @@
+    except UnkownQueryIdError:
+        return None
+
+
+def action_handler__poll_query(query_id):
+    """
+    Handler for the 'poll_query' action.
+
+    Returns the status of the query with the given `query_id`.
+    """
+    query_kind = _get_query_kind_for_query_id(query_id)
+    # TODO: we should probably be able to use the QueryStateMachine to determine
+    # whether the query already exists.
+    if query_kind is None:
+        return ZMQReply(status="error", msg=f"Unknown query id: '{query_id}'")
+    else:
+        q_state_machine = QueryStateMachine(Query.redis, query_id)
+        payload = {
+            "query_id": query_id,
+            "query_kind": query_kind,
+            "query_state": q_state_machine.current_query_state,
+        }
+        return ZMQReply(status="success", payload=payload)
+
+
+def action_handler__get_query_kind(query_id):
+    """
+    Handler for the 'get_query_kind' action.
+
+    Returns query kind of the query with the given `query_id`.
+    """
+    query_kind = _get_query_kind_for_query_id(query_id)
+    if query_kind is None:
+        error_msg = f"Unknown query id: '{query_id}'"
+        payload = {"query_id": query_id, "query_state": "awol"}
+        return ZMQReply(status="error", msg=error_msg, payload=payload)
+    else:
+        payload = {"query_id": query_id, "query_kind": query_kind}
+        return ZMQReply(status="success", payload=payload)
+
+
+def action_handler__get_query_params(query_id):
+    """
+    Handler for the 'get_query_params' action.
+
+    Returns query parameters of the query with the given `query_id`.
+    """
+    q_info_lookup = QueryInfoLookup(Query.redis)
+    try:
+        query_params = q_info_lookup.get_query_params(query_id)
+    except UnkownQueryIdError:
+        payload = {"query_id": query_id, "query_state": "awol"}
+        return ZMQReply(
+            status="error", msg=f"Unknown query id: '{query_id}'", payload=payload
+        )
+
+    payload = {"query_id": query_id, "query_params": query_params}
+    return ZMQReply(status="success", payload=payload)
+
+
+def action_handler__get_sql(query_id):
+    """
+    Handler for the 'get_sql' action.
+
+    Returns a SQL string which can be run against flowdb to obtain
+    the result of the query with given `query_id`.
+    """
+    # TODO: currently we can't use QueryStateMachine to determine whether
+    # the query_id belongs to a valid query object, so we need to check it
+    # manually. Would be good to add a QueryState.UNKNOWN so that we can
+    # avoid this separate treatment.
+    q_info_lookup = QueryInfoLookup(Query.redis)
+    if not q_info_lookup.query_is_known(query_id):
+        msg = f"Unknown query id: '{query_id}'"
+        payload = {"query_id": query_id, "query_state": "awol"}
+        return ZMQReply(status="error", msg=msg, payload=payload)
+
+    query_state = QueryStateMachine(Query.redis, query_id).current_query_state
+
+    if query_state == QueryState.COMPLETED:
```

Comment:
As noted, probably better to shove the messages into `QueryState` so this whole thing can vanish.

Reply:
Agreed. I'll open a follow-up issue.